### PR TITLE
[SW-1732] Implement shutdown hook to ensure H2O will go down on normal stop of Spark

### PR DESCRIPTION
**Pull request by jakubhava**
Depends on https://github.com/h2oai/h2o-3/pull/4110 (H2O 3.26.0.12)

No need for hacky stopping, tested on YARN in both internal & external backend

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/h2o/H2OContext.scala [hunk 1]
```diff
@@ -92,6 +93,10 @@ class H2OContext private(val sparkSession: SparkSession, conf: H2OConf) extends
     * otherwise it creates new H2O cluster living in Spark
     */
   def init(): H2OContext = {
+    ShutdownHookManager.addShutdownHook(10)(() => {
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: Would it be possible to add some comment explaining priority `10`?

Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: nit: WDYT about:
```
    ShutdownHookManager.addShutdownHook(10) { () =>
      logWarning("Spark shutdown hook called, stopping H2OContext!")
      stop(stopSparkContext = false, stopJvm = false)
    }
```

File: core/src/main/scala/org/apache/spark/h2o/H2OContext.scala [hunk 2]
```diff
@@ -282,21 +287,34 @@ class H2OContext private(val sparkSession: SparkSession, conf: H2OConf) extends
...
       logWarning("H2OContext is already stopped, this call has no effect anymore")
     }
   }
 
+  /** Stops H2O context.
+    *
+    * @param stopSparkContext stop also spark context
+    */
+  def stop(stopSparkContext: Boolean = false): Unit = {
+    stop(stopSparkContext, stopJvm = true)
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: Shouldn't we let Spark to stop JVM?

File: core/src/main/scala/org/apache/spark/h2o/H2OContext.scala (hunk 1 above)
Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: Sure, I need to provide comment in the code. We need to have higher ( lower number) priority then spark shutdown hooks so we stop H2O first and then let Spark stop itself

Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: I will make a constant for it with explanation

Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: 👍

File: core/src/main/scala/org/apache/spark/h2o/H2OContext.scala (hunk 2 above)
Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: In ideal case yes :) However the current stop method worked in a way that it always tried to stop JVM.

Once we have the client-less approach we can get happily rid of this, however I did not want to change behaviour in this release.

Comment (mn-mikke) [reply core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: ok, np.

File: core/src/main/scala/org/apache/spark/h2o/H2OContext.scala [hunk 3]
```diff
@@ -107,6 +109,10 @@ abstract class H2OContext private(val sparkSession: SparkSession, private val co
     * otherwise it creates new H2O cluster living in Spark
     */
   def init(): H2OContext = {
+    ShutdownHookManager.addShutdownHook(10) { () =>
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/H2OContext.scala]: Todo: explain why 10

**Comment by jakubhava** [comment]
@mn-mikke please have a look if it looks better now

**Pull request merged**